sdk/messaging/azservicebus/sender.go

```diff
@@ -77,83 +73,75 @@ func (s *Sender) NewMessageBatch(ctx context.Context, options ...MessageBatchOpt
 		}
 	}
 
-	return newMessageBatch(*opts.maxSizeInBytes), nil
+	return &MessageBatch{maxBytes: *opts.maxSizeInBytes}, nil
 }
 
 // SendMessage sends a message to a queue or topic.
 // Message can be a MessageBatch (created using `Sender.CreateMessageBatch`) or
 // a Message.
 func (s *Sender) SendMessage(ctx context.Context, message SendableMessage) error {
-	if s.linkState.Closed() {
-		return s.linkState.Err()
-	}
-
-	legacySender, err := s.createAmqpSender(ctx)
-
-	if err != nil {
-		return err
-	}
+	ctx, span := s.startProducerSpanFromContext(ctx, fmt.Sprintf(spanNameSendMessageFmt, message.MessageType()))
+	defer span.End()
 
-	amqpMessage, err := message.toAMQPMessage()
+	sender, _, _, _, err := s.links.Get(ctx)
 
 	if err != nil {
 		return err
 	}
 
-	return legacySender.SendAMQPMessage(ctx, amqpMessage)
+	return sender.Send(ctx, message.ToAMQPMessage())
 }
 
 // Close permanently closes the Sender.
 func (s *Sender) Close(ctx context.Context) error {
-	s.mu.Lock()
-	defer s.mu.Unlock()
-	defer s.linkState.Close()
-
-	var err error
-
-	if s.legacySender != nil {
-		err = s.legacySender.Close(ctx)
-		s.legacySender = nil
-	}
-
-	return err
+	return s.links.Close(ctx, true)
 }
 
-func (sender *Sender) createAmqpSender(ctx context.Context) (internal.LegacySender, error) {
-	sender.mu.Lock()
-	defer sender.mu.Unlock()
+func (sender *Sender) createSenderLink(ctx context.Context, session internal.AMQPSession) (internal.AMQPSenderCloser, internal.AMQPReceiverCloser, error) {
+	amqpSender, err := session.NewSender(
+		amqp.LinkSenderSettle(amqp.ModeMixed),
+		amqp.LinkReceiverSettle(amqp.ModeFirst),
+		amqp.LinkTargetAddress(sender.queueOrTopic))
 
-	if sender.legacySender != nil {
-		return sender.legacySender, nil
-	}
-
-	// TODO: allow passing in relevant options if needed
-	legacySender, err := sender.nsCreateLegacySender(ctx, sender.config.queueOrTopic)
-
-	if err != nil {
-		return nil, err
-	}
-
-	sender.legacySender = legacySender
-	return legacySender, nil
+	return amqpSender, nil, err
 }
 
-// ie: `*internal.Namespace`
-type legacySenderNamespace interface {
-	NewLegacySender(ctx context.Context, entityPath string) (internal.LegacySender, error)
-}
-
-func newSender(ns legacySenderNamespace, queueOrTopic string) (*Sender, error) {
+func newSender(ns *internal.Namespace, queueOrTopic string) (*Sender, error) {
 	sender := &Sender{
-		config: struct {
-			queueOrTopic string
-		}{
-			queueOrTopic: queueOrTopic,
-		},
-		mu:        &sync.Mutex{},
-		linkState: newLinkState(context.Background(), errClosed{link: "sender"}),
+		queueOrTopic: queueOrTopic,
 	}
 
-	sender.nsCreateLegacySender = ns.NewLegacySender
+	sender.links = ns.NewAMQPLinks(queueOrTopic, sender.createSenderLink)
 	return sender, nil
 }
+
+// handleAMQPError is called internally when an event has failed to send so we
```

Comment:
Can this comment block be removed or is it waiting for something?

Reply:
Gone. I can always get it back from history later if needed.